# Require that member overloads be consecutive (`adjacent-overload-signatures`)

Grouping overloaded members together can improve readability of the code.

## Rule Details

This rule aims to standardize the way overloaded members are organized.

<!--tabs-->

### ‚ùå Incorrect

```ts
declare namespace Foo {
  export function bar(s: string): void;
  export function bar(n: number): void;
  export function bar(): void;
  export function bar(sn: string | number): void;
}

type Foo = {
  bar(s: string): void;
  bar(n: number): void;
  bar(): void;
  bar(sn: string | number): void;
};

interface Foo {
  bar(s: string): void;
  bar(n: number): void;
  bar(): void;
  bar(sn: string | number): void;
}

class Foo {
  bar(s: string): void;
  bar(n: number): void;
  bar(): void {}
  bar(sn: string | number): void {}
}

export function bar(s: string): void;
export function bar(n: number): void;
export function bar(): void;
export function bar(sn: string | number): void;
```

### ‚úÖ Correct

```ts
declare namespace Foo {
  export function bar(s: string): void;
  export function bar(n: number): void;
  export function bar(sn: string | number): void;
  export function bar(): void;
}

type Foo = {
  bar(s: string): void;
  bar(n: number): void;
  bar(sn: string | number): void;
  bar(): void;
};

interface Foo {
  bar(s: string): void;
  bar(n: number): void;
  bar(sn: string | number): void;
  bar(): void;
}

class Foo {
  bar(s: string): void;
  bar(n: number): void;
  bar(sn: string | number): void {}
  bar(): void {}
}

export function bar(): void;
export function bar(s: string): void;
export function bar(n: number): void;
export function bar(sn: string | number): void;
```

## When Not To Use It

If you don't care about the general structure of the code, then you will not need this rule.

## Related To

- TSLint: [adjacent-overload-signatures](https://palantir.github.io/tslint/rules/adjacent-overload-signatures/)

## Attributes

- [x] ‚úÖ Recommended
- [ ] üîß Fixable
- [ ] üí≠ Requires type information
