# Disallow extra non-null assertion (`no-extra-non-null-assertion`)

## Rule Details

Examples of code for this rule:

<!--tabs-->

### ❌ Incorrect

```ts
const bar: { bar: number } | null = null;
const bar = bar!!!.bar;
```

```ts
function bar(bar: number | undefined) {
  const bar: number = bar!!!;
}
```

```ts
function bar(bar?: { n: number }) {
  return bar!?.n;
}
```

### ✅ Correct

```ts
const bar: { bar: number } | null = null;
const bar = bar!.bar;
```

```ts
function bar(bar: number | undefined) {
  const bar: number = bar!;
}
```

```ts
function bar(bar?: { n: number }) {
  return bar?.n;
}
```

## How to Use

```json
{
  "@typescript-eslint/no-extra-non-null-assertion": ["error"]
}
```

## Attributes

- [x] ✅ Recommended
- [x] 🔧 Fixable
- [ ] 💭 Requires type information
