# Warns if a type assertion does not change the type of an expression (`no-unnecessary-type-assertion`)

This rule prohibits using a type assertion that does not change the type of an expression.

## Rule Details

This rule aims to prevent unnecessary type assertions.

Examples of code for this rule:

<!--tabs-->

### ‚ùå Incorrect

```ts
const bar = 3;
const bar = bar!;
```

```ts
const bar = <3>3;
```

```ts
type Foo = 3;
const bar = <Foo>3;
```

```ts
type Foo = 3;
const bar = 3 as Foo;
```

```ts
function bar(x: number): number {
  return x!; // unnecessary non-null
}
```

### ‚úÖ Correct

```ts
const bar = <number>3;
```

```ts
const bar = 3 as number;
```

```ts
const bar = 'bar' as const;
```

```ts
function bar(x: number | undefined): number {
  return x!;
}
```

## Options

This rule optionally takes an object with a single property `typesToIgnore`, which can be set to a list of type names to ignore.

For example, with `@typescript-eslint/no-unnecessary-type-assertion: ["error", { typesToIgnore: ['Foo'] }]`, the following is **correct** code":

```ts
type Foo = 3;
const bar: Foo = 3;
```

## When Not To Use It

If you don't care about having no-op type assertions in your code, then you can turn off this rule.

## Related To

- TSLint: [`no-unnecessary-type-assertion`](https://palantir.github.io/tslint/rules/no-unnecessary-type-assertion/)

## Attributes

- [x] ‚úÖ Recommended
- [x] üîß Fixable
- [x] üí≠ Requires type information
