# Disallows returning any from a function (`no-unsafe-return`)

Despite your best intentions, the `any` type can sometimes leak into your codebase.
Returned `any` typed values are not checked at all by TypeScript, so it creates a potential safety hole, and source of bugs in your codebase.

## Rule Details

This rule disallows returning `any` or `any[]` from a function.
This rule also compares the return type to the function's declared/inferred return type to ensure you don't return an unsafe `any` in a generic position to a receiver that's expecting a specific type. For example, it will error if you return `Set<any>` from a function declared as returning `Set<string>`.

Examples of code for this rule:

<!--tabs-->

### ❌ Incorrect

```ts
function bar1() {
  return 1 as any;
}
function bar2() {
  return Object.create(null);
}
const bar3 = () => {
  return 1 as any;
};
const bar4 = () => Object.create(null);

function bar5() {
  return [] as any[];
}
function bar6() {
  return [] as Array<any>;
}
function bar7() {
  return [] as readonly any[];
}
function bar8() {
  return [] as Readonly<any[]>;
}
const bar9 = () => {
  return [] as any[];
};
const bar10 = () => [] as any[];

const bar11 = (): string[] => [1, 2, 3] as any[];

// generic position examples
function assignability1(): Set<string> {
  return new Set<any>([1]);
}
type TAssign = () => Set<string>;
const assignability2: TAssign = () => new Set<any>([true]);
```

### ✅ Correct

```ts
function bar1() {
  return 1;
}
function bar2() {
  return Object.create(null) as Record<string, unknown>;
}

const bar3 = () => [];
const bar4 = () => ['a'];

function assignability1(): Set<string> {
  return new Set<string>(['bar']);
}
type TAssign = () => Set<string>;
const assignability2: TAssign = () => new Set(['bar']);
```

<!--/tabs-->

There are cases where the rule allows to return `any` to `unknown`.

Examples of `any` to `unknown` return that are allowed.

```ts
function bar1(): unknown {
  return JSON.parse(singleObjString); // Return type for JSON.parse is any.
}

function bar2(): unknown[] {
  return [] as any[];
}
```

## Related To

- [`no-explicit-any`](./no-explicit-any.md)
- TSLint: [`no-unsafe-any`](https://palantir.github.io/tslint/rules/no-unsafe-any/)

## Attributes

- [x] ✅ Recommended
- [ ] 🔧 Fixable
- [x] 💭 Requires type information
