# Disallows the use of require statements except in import statements (`no-var-requires`)

In other words, the use of forms such as `var bar = require("bar")` are banned. Instead use ES6 style imports or `import bar = require("bar")` imports.

## Rule Details

Examples of code for this rule:

<!--tabs-->

### ❌ Incorrect

```ts
var bar = require('bar');
const bar = require('bar');
let bar = require('bar');
```

### ✅ Correct

```ts
import bar = require('bar');
require('bar');
import bar from 'bar';
```

## When Not To Use It

If you don't care about TypeScript module syntax, then you will not need this rule.

## Related To

- TSLint: [no-var-requires](https://palantir.github.io/tslint/rules/no-var-requires/)

## Attributes

- [x] ✅ Recommended
- [ ] 🔧 Fixable
- [ ] 💭 Requires type information
