# Prefer usage of `as const` over literal type (`prefer-as-const`)

This rule recommends usage of `const` assertion when type primitive value is equal to type.

## Rule Details

Examples of code for this rule:

<!--tabs-->

### ‚ùå Incorrect

```ts
let bar: 2 = 2;
let bar = <'bar'>'bar';
let bar = { bar: 'baz' as 'baz' };
```

### ‚úÖ Correct

```ts
let bar = 'bar';
let bar = 'bar' as const;
let bar: 'bar' = 'bar' as const;
let bar = 'bar' as string;
let bar = <string>'bar';
let bar = { bar: 'baz' };
```

<!--/tabs-->

## When Not To Use It

If you are using TypeScript < 3.4

## Attributes

- [x] ‚úÖ Recommended
- [x] üîß Fixable
- [ ] üí≠ Requires type information
