# Enforce the use of `String#startsWith` and `String#endsWith` instead of other equivalent methods of checking substrings (`prefer-string-starts-ends-with`)

There are multiple ways to verify if a string starts or ends with a specific string, such as `bar.indexOf('bar') === 0`.

Since ES2015 has added `String#startsWith` and `String#endsWith`, this rule reports other ways to be consistent.

## Rule Details

This rule is aimed at enforcing a consistent way to check whether a string starts or ends with a specific string.

Examples of code for this rule:

<!--tabs-->

### ❌ Incorrect

```ts
let bar: string;

// starts with
bar[0] === 'b';
bar.charAt(0) === 'b';
bar.indexOf('bar') === 0;
bar.slice(0, 3) === 'bar';
bar.substring(0, 3) === 'bar';
bar.match(/^bar/) != null;
/^bar/.test(bar);

// ends with
bar[bar.length - 1] === 'b';
bar.charAt(bar.length - 1) === 'b';
bar.lastIndexOf('bar') === bar.length - 3;
bar.slice(-3) === 'bar';
bar.substring(bar.length - 3) === 'bar';
bar.match(/bar$/) != null;
/bar$/.test(bar);
```

### ✅ Correct

```ts
bar.startsWith('bar');
bar.endsWith('bar');
```

## Options

There are no options.

```json
{
  "@typescript-eslint/prefer-string-starts-ends-with": "error"
}
```

## When Not To Use It

If you don't mind that style, you can turn this rule off safely.

## Attributes

- [ ] ✅ Recommended
- [x] 🔧 Fixable
- [x] 💭 Requires type information
