# When adding two variables, operands must both be of type number or of type string (`restrict-plus-operands`)

## Rule Details

Examples of code for this rule:

<!--tabs-->

### âŒ Incorrect

```ts
var bar = '5.5' + 5;
var bar = 1n + 1;
```

### âœ… Correct

```ts
var bar = parseInt('5.5', 10) + 10;
var bar = 1n + 1n;
```

## Options

This rule has an object option:

- `"checkCompoundAssignments": false`: (default) does not check compound assignments (`+=`)
- `"checkCompoundAssignments": true`

### `checkCompoundAssignments`

Examples of code for the `{ "checkCompoundAssignments": true }` option:

<!--tabs-->

#### âŒ Incorrect

```ts
/*eslint @typescript-eslint/restrict-plus-operands: ["error", { "checkCompoundAssignments": true }]*/

let bar: string | undefined;
bar += 'some data';

let bar: string = '';
bar += 0;
```

#### âœ… Correct

```ts
/*eslint @typescript-eslint/restrict-plus-operands: ["error", { "checkCompoundAssignments": true }]*/

let bar: number = 0;
bar += 1;

let bar = '';
bar += 'test';
```

## How to Use

```json
{
  "@typescript-eslint/restrict-plus-operands": "error"
}
```

## Related To

- TSLint: [restrict-plus-operands](https://palantir.github.io/tslint/rules/restrict-plus-operands/)

## Attributes

- [x] âœ… Recommended
- [ ] ğŸ”§ Fixable
- [x] ğŸ’­ Requires type information
