---
description: 'Require that function overload signatures be consecutive.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/adjacent-overload-signatures** for documentation.

Function overload signatures represent multiple ways a function can be called, potentially with different return types.
It's typical for an interface or type alias describing a function to place all overload signatures next to each other.
If Signatures placed elsewhere in the type are easier to be missed by future developers reading the code.

## Examples

<!--tabs-->

### ‚ùå Incorrect

```ts
declare namespace Foo {
  export function bar(s: string): void;
  export function bar(n: number): void;
  export function bar(): void;
  export function bar(sn: string | number): void;
}

type Foo = {
  bar(s: string): void;
  bar(n: number): void;
  bar(): void;
  bar(sn: string | number): void;
};

interface Foo {
  bar(s: string): void;
  bar(n: number): void;
  bar(): void;
  bar(sn: string | number): void;
}

class Foo {
  bar(s: string): void;
  bar(n: number): void;
  bar(): void {}
  bar(sn: string | number): void {}
}

export function bar(s: string): void;
export function bar(n: number): void;
export function bar(): void;
export function bar(sn: string | number): void;
```

### ‚úÖ Correct

```ts
declare namespace Foo {
  export function bar(s: string): void;
  export function bar(n: number): void;
  export function bar(sn: string | number): void;
  export function bar(): void;
}

type Foo = {
  bar(s: string): void;
  bar(n: number): void;
  bar(sn: string | number): void;
  bar(): void;
};

interface Foo {
  bar(s: string): void;
  bar(n: number): void;
  bar(sn: string | number): void;
  bar(): void;
}

class Foo {
  bar(s: string): void;
  bar(n: number): void;
  bar(sn: string | number): void {}
  bar(): void {}
}

export function bar(): void;
export function bar(s: string): void;
export function bar(n: number): void;
export function bar(sn: string | number): void;
```

## When Not To Use It

If you don't care about the general structure of the code, then you will not need this rule.
