---
description: 'Disallow non-null assertions in the left operand of a nullish coalescing operator.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/no-non-null-asserted-nullish-coalescing** for documentation.

The `??` nullish coalescing runtime operator allows providing a default value when dealing with `null` or `undefined`.
Using a `!` non-null assertion type operator in the left operand of a nullish coalescing operator is redundant, and likely a sign of programmer error or confusion over the two operators.

## Examples

<!--tabs-->

### ‚ùå Incorrect

```ts
bar! ?? bar;
bar.bazz! ?? bar;
bar!.bazz! ?? bar;
bar()! ?? bar;

let x!: string;
x! ?? '';

let x: string;
x = bar();
x! ?? '';
```

### ‚úÖ Correct

```ts
bar ?? bar;
bar ?? bar!;
bar!.bazz ?? bar;
bar!.bazz ?? bar!;
bar() ?? bar;

// This is considered correct code because there's no way for the user to satisfy it.
let x: string;
x! ?? '';
```

## Further Reading

- [TypeScript 3.7 Release Notes](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html)
- [Nullish Coalescing Proposal](https://github.com/tc39/proposal-nullish-coalescing)
