---
description: 'Disallow returning a value with type `any` from a function.'
---

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-unsafe-return** for documentation.

The `any` type in TypeScript is a dangerous "escape hatch" from the type system.
Using `any` disables many type checking rules and is generally best used only as a last resort or when prototyping code.

Despite your best intentions, the `any` type can sometimes leak into your codebase.
Returning an an `any`-typed value from a function creates a potential type safety hole and source of bugs in your codebase.

This rule disallows returning `any` or `any[]` from a function.

This rule also compares generic type argument types to ensure you don't return an unsafe `any` in a generic position to a function that's expecting a specific type.
For example, it will error if you return `Set<any>` from a function declared as returning `Set<string>`.

## Examples

<!--tabs-->

### âŒ Incorrect

```ts
function bar1() {
  return 1 as any;
}
function bar2() {
  return Object.create(null);
}
const bar3 = () => {
  return 1 as any;
};
const bar4 = () => Object.create(null);

function bar5() {
  return [] as any[];
}
function bar6() {
  return [] as Array<any>;
}
function bar7() {
  return [] as readonly any[];
}
function bar8() {
  return [] as Readonly<any[]>;
}
const bar9 = () => {
  return [] as any[];
};
const bar10 = () => [] as any[];

const bar11 = (): string[] => [1, 2, 3] as any[];

// generic position examples
function assignability1(): Set<string> {
  return new Set<any>([1]);
}
type TAssign = () => Set<string>;
const assignability2: TAssign = () => new Set<any>([true]);
```

### âœ… Correct

```ts
function bar1() {
  return 1;
}
function bar2() {
  return Object.create(null) as Record<string, unknown>;
}

const bar3 = () => [];
const bar4 = () => ['a'];

function assignability1(): Set<string> {
  return new Set<string>(['bar']);
}
type TAssign = () => Set<string>;
const assignability2: TAssign = () => new Set(['bar']);
```

<!--/tabs-->

There are cases where the rule allows to return `any` to `unknown`.

Examples of `any` to `unknown` return that are allowed:

```ts
function bar1(): unknown {
  return JSON.parse(singleObjString); // Return type for JSON.parse is any.
}

function bar2(): unknown[] {
  return [] as any[];
}
```

## Related To

- [`no-explicit-any`](./no-explicit-any.md)
