---
description: 'Disallow `require` statements except in import statements.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/no-var-requires** for documentation.

In other words, the use of forms such as `var bar = require("bar")` are banned. Instead use ES6 style imports or `import bar = require("bar")` imports.

## Examples

<!--tabs-->

### ‚ùå Incorrect

```ts
var bar = require('bar');
const bar = require('bar');
let bar = require('bar');
```

### ‚úÖ Correct

```ts
import bar = require('bar');
require('bar');
import bar from 'bar';
```

## When Not To Use It

If you don't care about using newer module syntax, then you will not need this rule.

## Related To

- [`no-require-imports`](./no-require-imports.md)
