---
description: 'Enforce using `String#startsWith` and `String#endsWith` over other equivalent methods of checking substrings.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/prefer-string-starts-ends-with** for documentation.

There are multiple ways to verify if a string starts or ends with a specific string, such as `bar.indexOf('bar') === 0`.
As of ES2015, the most common way in JavaScript is to use `String#startsWith` and `String#endsWith`.
Keeping to those methods consistently helps with code readability.

This rule reports when a string method can be replaced safely with `String#startsWith` or `String#endsWith`.

## Examples

<!--tabs-->

### ‚ùå Incorrect

```ts
declare const bar: string;

// starts with
bar[0] === 'b';
bar.charAt(0) === 'b';
bar.indexOf('bar') === 0;
bar.slice(0, 3) === 'bar';
bar.substring(0, 3) === 'bar';
bar.match(/^bar/) != null;
/^bar/.test(bar);

// ends with
bar[bar.length - 1] === 'b';
bar.charAt(bar.length - 1) === 'b';
bar.lastIndexOf('bar') === bar.length - 3;
bar.slice(-3) === 'bar';
bar.substring(bar.length - 3) === 'bar';
bar.match(/bar$/) != null;
/bar$/.test(bar);
```

### ‚úÖ Correct

```ts
declare const bar: string;

// starts with
bar.startsWith('bar');

// ends with
bar.endsWith('bar');
```

## When Not To Use It

If you don't mind that style, you can turn this rule off safely.
